### DISCUSSION TITLE
How would I parse markdown `h1` into `rehype-document` title property

### DISCUSSION THREAD

**Discussion by phortuin** [remarkjs/remark#525] [Q&A] [answered]
I’m trying to find a way to parse a markdown `h1` into a HTML document’s `<title>` element (not sure if this belongs here or in Rehype discussions). Example:

```markdown
# My document

This is my document
```

Index.js should be something like this:

```js
const processor = unified()
  .use(markdown)
  .use(/** magic plugin */)
  .use(remark2rehype)
  .use(doc, {
    title: /** magic property */
  })
  .use(stringify)
```

Output:

```html
<!doctype html>
<html>
<head>
<title>My document</title>
</head>
<body>
<h1>My document</h1>
<p>This is my document</p>
</body>
</html>
```

The `magic plugin` and `magic property` parts above are ofcourse where I'm stuck. I'm not sure how I would add a thing to the AST (I've looked at `remark-frontmatter` for inspiration) and take stuff out of it and pass it as a property on the `rehype-document` (`doc`) plugin options. I’ve looked at several plugins and read documentation but can’t find the missing 'link. Any help would be greatly appreciated!

**Comment by wooorm** [comment, marked as answer]
this is more of a rehype thing indeed, as we’re not worried about markdown in this case.

Typically, in unified, everything can be done with plugins: ast transforms. To solve this, one could write a custom plugin to find the `h1`, and the `title`, and when both exist, get the text content of the heading and inject it into the title. For inspiration, see this pseudocode:

```js
var select = require('hast-util-select').select
var fromString = require('hast-util-from-string')
var toString = require('hast-util-to-string')

.use(myPlugin)

function myPlugin() {
  return myTransformer
  function myTransformer(tree) {
    var h1 = select('h1', tree)
    var title = select('title', tree)
    if (h1 && title) {
      fromString(title, toString(h1))
    }
  }
}
```

**Reply by phortuin** [reply]
Awesome. Thanks for the lightning fast response, and to my surprise this was much, much easier (and more straightforward) than I expected. For future reference:

```js
const select = require('hast-util-select').select
const fromString = require('hast-util-from-string')
const toString = require('hast-util-to-string')

module.exports = h1ToTitle

function h1ToTitle() {
    return transformer
}

function transformer(tree) {
    const h1 = select('h1', tree)
    const title = select('title', tree)
    if (h1 && title) {
        fromString(title, toString(h1))
    }
}
```

Its only caveat is that we’re now forced to set a `title` property with `rehype-document`, which actually works out fine when set to 'Untitled':

```js
const h1ToTitle = require('./lib/rehype-h1-to-title')

const processor = unified()
    .use(markdown)
    .use(remark2rehype)
    .use(doc, {
        title: 'Untitled',
    })
    .use(h1ToTitle)
    .use(stringify)
```
